### ISSUE TITLE
Episodes/Movies that have been fully buffered do not play or scrub

### ISSUE THREAD

**Issue by mattjis** [PopcornTimeTV/PopcornTimeTV#211]
If I do then when I click play again only the audio plays, leaving the video frozen :(

I have the latest beta 0.2.0-5

**Comment by FlaviaBondJamesBondsHusband** [comment]
The latest beta is 0.2.0-6 not 0.2.0-5. Please try again with 0.2.0-6

**Comment by mattjis** [comment]
Does the built in updater not work? "No Updates Available"

I'll download the latest version now though and let you know! Thanks

**Comment by FlaviaBondJamesBondsHusband** [comment]
@mattjis it doesn't work in versions below and including `0.2.0-5` due to a bug which has now been fixed.

**Comment by mattjis** [comment]
@FlaviaBondJamesBondsHusband I just tried to get it but somehow ended up with 0.2.0-5 again :( 

This time though when I opened it, it said "Beta 0.2.0-6 of Popcorn Time is now available" and gives me 3 options:
1. "Next Time"
2. "Skip this version"
3. "OK"

None of them seem to do anything. Will I have to try again or is there a way around?

I got this version by "swift install.swift" and typing in "0.2.0-6" so I'm confused as to why I'm only on 0.2.0-5

**Comment by FlaviaBondJamesBondsHusband** [comment]
@mattjis i dont think the build number was updated for the 0.2.0-6 version but if scrubbing works, you are definitely on the latest version. Regarding the update manager, we cannot actually update the app without the user going onto their computer and re-running the install script and downloading the latest version so it is more of an update Notifier. Could you try to see if scrubbing is working?

**Comment by mattjis** [comment]
@FlaviaBondJamesBondsHusband ohh okay I understand!

Nope doesn't seem to work. I can pause fine but when I try to resume both the audio & video stay frozen indefinitely - with a spiral loading animation next to the time. 

I have to go back to the menu and start over / redownload (from beginning) to get it to work again :(

**Comment by FlaviaBondJamesBondsHusband** [comment]
The buffering icon will stay there until the movie is done buffering. Torrenting is different to normal HLS and will therefore take longer to load. Just be patient and the movie will begin playing again.

**Comment by mattjis** [comment]
I've been waiting around 8 minutes since pausing and then resuming. The screen is still frozen with the spiral animation twirling. Does it have to download the whole movie / episode before I can continue watching?

**Comment by FlaviaBondJamesBondsHusband** [comment]
@mattjis 

> The video was also only paused at 5 seconds in

We are using `MobileVLCKit 2.7.9` for all the movie and episode streaming. 2.7.9 has not been properly updated for tvOS/iOS 10. They do offer an alpha version `3.0.0` but that is too unstable to use for this project yet. Because of this, at the start of every movie there is an issue where the movie will stop playing about 5-10 seconds in. To combat this you can just rewind a few seconds and start playing again and it should work as expected. Unfortunately there is nothing we can do until VLC release a stable version of `3.0.0`.


> Does it have to download the whole movie / episode before I can "unpause"?

No, it will skip the torrent downloading up to the place you have selected and will begin from there. If you are watching a movie with low seeds, it will take considerably longer, nothing we can do about that unfortunately. Once the torrent is done buffering, it will automatically resume playback.

**Comment by mattjis** [comment]
Ohh I think I've explained myself incorrectly. The video didn't freeze or pause on it's own.

I purposely paused it 5 seconds in after it began playing (the latest episode of modern family with hundreds of seeds) and once I had paused it I tried to resume it again (This is the problem i'm facing). It would not and still hasn't resumed playing 20 minutes later. 

My internet is quite fast and in this time I could have easily downloaded the full episode 5 times over. 

Basically all I'm saying is I've paused it and can now no longer resume it. The only way I've figured out how to fix this is to click menu and start downloading it from the beginning again.

No amount of scrubbing back or forward seems to help :( I hope I make sense this time

**Comment by FlaviaBondJamesBondsHusband** [comment]
@mattjis What is the exact episode/season and what quality are you trying to watch it in?

**Comment by mattjis** [comment]
@FlaviaBondJamesBondsHusband Modern Family S08E08 at 720P

**Comment by FlaviaBondJamesBondsHusband** [comment]
@mattjis Episode 7 and 8 in season 8 have 0 seeds or peers for either quality. 

>  If you are watching a movie with low seeds, it will take considerably longer, nothing we can do about that unfortunately.

**Comment by mattjis** [comment]
@FlaviaBondJamesBondsHusband The episode came out less than a week ago & is from one of the most popular TV shows in the world so I highly doubt that. 

When I conducted this test myself it even said it had connected to 20+ seeds before switching to the player view. Check for yourself: http://1337x.to/search/modern+family+s08e08/1/

**Comment by FlaviaBondJamesBondsHusband** [comment]
@mattjis 

> The episode came out less than a week ago & is from one of the most popular TV shows in the world so I highly doubt that.

The fact that it came out last week is reason enough as to it having low seed and peer count!

> When I conducted this test myself it even said it had connected to 20+ seeds before switching to the player view. Check for yourself: http://1337x.to/search/modern+family+s08e08/1/

We don't take torrents from there. I'm literally looking at the api now and it says there are 0 seeds and 0 peers. 
![screen shot 2016-12-04 at 18 37 26](https://cloud.githubusercontent.com/assets/20926198/20868353/8abe5338-ba51-11e6-97f8-010909c75793.png)

If you could test episode 6 to confirm what i was saying please?

**Comment by mattjis** [comment]
@FlaviaBondJamesBondsHusband 

> The fact that it came out last week is reason enough as to it having low seed and peer count!

In my experience tv show torrents are always healthiest within the first week of being uploaded. I can guarantee that that seeder count is only going downhill from here onward.

And with that link i sent, you can clearly see the 2 eztv torrents with: 1898 seeds & 568 seeds.

I also checked those magnet links from the image you attached and they match the 2 eztv links I was referencing (1898 seeds & 568 seeds) which means you might have other problems that are showing those incorrect seeder counts.

Episode 6 has the same issue as well for me :( I can record it if you think that it would be helpful as it's a real pain.

**Comment by delut** [comment]
I'm having this same issue (I think). 

Every time I press pause, and then play again the video freezes and I can't resume playback. The UI is properly updated and reacts correctly to the play/pause button press. But the video won't play again. No matter at what point of the show I am, whether or not the episode was fully downloaded.

This happened to me with every episode and movie I've tried.

Its hard to explain, maybe attaching a short video would help?

**Comment by FlaviaBondJamesBondsHusband** [comment]
@delut i can confirm im having this issue with episodes, but cannot reproduce on movies. What movie have you tried?

**Comment by delut** [comment]
@FlaviaBondJamesBondsHusband it happens with every single movie/show. I've just tried it with Star Wars: The Force Awakens in 720p and it happens.

I've been investigating this issue, and I think there's something happening when VLC reads the file thats being written by the streamer.

The play/pause issue comes because, when you pause and later play at the same position, the playback time is calculated and later set (even if it doesn't change). In `func clickGesture(_ gesture: VLCSiriRemoteGestureRecognizer)` in the `PCTPlayerViewController+tvOS.swift` file (lines 98 and 99).

**Setting the `mediaplayer.time` property on the player is what breaks things**. I'll investigate further on why things break later. But Xcode's console throws the following logs whenever the property is set:

```
[ERROR] Error while writing to socket 111: Broken pipe (32)
2016-12-08 17:12:17.847396 PopcornTime[316:92589] new startPiece: 5
[ERROR] Error while writing to socket 94: Broken pipe (32)
2016-12-08 17:12:20.823537 PopcornTime[316:92626] new startPiece: 8
```
Scrubbing does work most of the times with the simulator, but not once when working with an Apple TV. I think there might be some kind of issue in the way VLC reads the file while its being written.

Finally I'd like to point out that there are two separate issues:

1. The player/streamer breaks whenever the property `mediaplayer.time` is set and this happens almost exclusively on the device (it happens all the time). Whenever this happens the video playback can't be resumed. The video does play until `mediaplayer.time` is manually updated.
2. Whenever a file was fully buffered, it can't be played. It doesn't start. Clearing the cache from settings makes it work again.

**Comment by FlaviaBondJamesBondsHusband** [comment]
> The play/pause issue comes because, when you pause and later play at the same position, the playback time is calculated and later set (even if it doesn't change). In func clickGesture(_ gesture: VLCSiriRemoteGestureRecognizer) in the PCTPlayerViewController+tvOS.swift file (lines 98 and 99).

Whats the "play/pause issue"? Playing and pausing work fine. I fail to see how setting the time even if it doesn't change is an issue? If the user want to cancel scrubbing, they press the menu button, otherwise the time is set. 

> [ERROR] Error while writing to socket 111: Broken pipe (32)
2016-12-08 17:12:17.847396 PopcornTime[316:92589] new startPiece: 5
[ERROR] Error while writing to socket 94: Broken pipe (32)
2016-12-08 17:12:20.823537 PopcornTime[316:92626] new startPiece: 8

These 'errors' are not actually errors. That is just GCDWebserver spewing out stuff to the console. It does this all the time and it doesn't mean anything. `new startPiece: 8` just signifies that the movie will start buffering from piece 8, the piece the user wants to play, instead of starting from piece 0 and buffering up to piece 8 even if the user doesn't want pieces 0-7.

> Scrubbing does work most of the times with the simulator, but not once when working with an Apple TV

Scrubbing works 100% of the time on the simulator except for when the movie/episode is fully buffered. The issue is strictly confined to devices which I was not aware of as we usually only test on simulator.

We are investigating the issue further now and hope to have a fix by release 2.1.0

**Comment by delut** [comment]
This is the play/pause bug. Whenever you pause the spinner appears and you can't play again. Some parts of the UI become unresponsive.

[Gif reproducing issue](http://i.imgur.com/Rb6eA18.gif)

I sort of solved the issue. Updating _TVVLCKit_ to the latest 3.0.0 nightly makes everything work. It was indeed a VLC problem. I've read earlier that you wrote somewhere that these builds weren't stable enough yet. But then, this is a solved issue. Whenever they release the final 3.0.0 release everything should work.

I can confirm then that its solved with version `TVVLCKit-unstable-3.0.0a24` found here http://download.videolan.org/cocoapods/unstable/.

**Comment by hvluu** [comment]
i confirmed that updating the TVVLCKit did fix the playing/pausing bug

**Comment by FlaviaBondJamesBondsHusband** [comment]
@delut @hvluu thanks for your work. We have rectified the "play/pause" scrubbing issue. However, I shall leave this issue open until we fix the scrubbing/playing when fully buffered bug

**Comment by mattjis** [comment]
I can also confirm that it now works as expected on version 0.2.0-7. Thanks for looking in to this @FlaviaBondJamesBondsHusband :)

**Comment by derickito** [comment]
I'm on version 0.2.0-7 and this issue still persist. It can be reproduced on movies or show episodes by waiting for the entire thing to load like title says. After that I can't rewind with the remote using the 30-second rewind (clicking on the left side of the remote).

**Comment by FlaviaBondJamesBondsHusband** [comment]
Issue should be resolved now
